fv3core/stencils/divergence_damping.py

```diff
@@ -19,60 +20,56 @@ def damp_tmp(q, da_min_c, d2_bg, dddmp):
     return damp
 
 
-def ptc_computation(
+def get_delpc(
     u: FloatField,
-    va: FloatField,
-    vc: FloatField,
-    cosa_v: FloatFieldIJ,
-    sina_v: FloatFieldIJ,
-    dyc: FloatFieldIJ,
-    sin_sg2: FloatFieldIJ,
-    sin_sg4: FloatFieldIJ,
-    ptc: FloatField,
-):
-    """computation of pct"""
-    from __externals__ import j_end, j_start
-
-    with computation(PARALLEL), interval(...):
-        ptc = (u - 0.5 * (va[0, -1, 0] + va) * cosa_v) * dyc * sina_v
-        with horizontal(region[:, j_start], region[:, j_end + 1]):
-            ptc = u * dyc * sin_sg4[0, -1] if vc > 0 else u * dyc * sin_sg2
-
-
-def vorticity_computation(
     v: FloatField,
     ua: FloatField,
+    va: FloatField,
     cosa_u: FloatFieldIJ,
     sina_u: FloatFieldIJ,
     dxc: FloatFieldIJ,
-    vort: FloatField,
+    dyc: FloatFieldIJ,
     uc: FloatField,
-    sin_sg3: FloatFieldIJ,
+    vc: FloatField,
     sin_sg1: FloatFieldIJ,
-):
-    """computation of the vorticity"""
-    from __externals__ import i_end, i_start
-
-    with computation(PARALLEL), interval(...):
-        vort = (v - 0.5 * (ua[-1, 0, 0] + ua) * cosa_u) * dxc * sina_u
-        with horizontal(region[i_start, :], region[i_end + 1, :]):
-            vort = v * dxc * sin_sg3[-1, 0] if uc > 0 else v * dxc * sin_sg1
-
-
-def delpc_computation(
-    ptc: FloatField,
+    sin_sg2: FloatFieldIJ,
+    sin_sg3: FloatFieldIJ,
+    sin_sg4: FloatFieldIJ,
+    cosa_v: FloatFieldIJ,
+    sina_v: FloatFieldIJ,
     rarea_c: FloatFieldIJ,
     delpc: FloatField,
-    vort: FloatField,
+    u_contra_dyc: FloatField,
+    v_contra_dxc: FloatField,
 ):
     from __externals__ import i_end, i_start, j_end, j_start
 
     with computation(PARALLEL), interval(...):
-        delpc = vort[0, -1, 0] - vort + ptc[-1, 0, 0] - ptc
+        # TODO: why does uc_from_ua sometimes have different sign than uc?
+        uc_from_ua = 0.5 * (ua[-1, 0, 0] + ua)
+        # TODO: why do we use uc_from_ua and not just uc?
+        v_contra = contravariant(v, uc_from_ua, cosa_u, sina_u)
+        with horizontal(region[i_start, :], region[i_end + 1, :]):
+            v_contra = v * sin_sg3[-1, 0] if uc > 0 else v * sin_sg1
+        v_contra_dxc = v_contra * dxc
+
+        vc_from_va = 0.5 * (va[0, -1, 0] + va)
+        u_contra = contravariant(u, vc_from_va, cosa_v, sina_v)
+        u_contra = (u - 0.5 * (va[0, -1, 0] + va) * cosa_v) * sina_v
+        with horizontal(region[:, j_start], region[:, j_end + 1]):
+            u_contra = u * sin_sg4[0, -1] if vc > 0 else u * sin_sg2
+        u_contra_dyc = u_contra * dyc
+
+        delpc = (
+            v_contra_dxc[0, -1, 0]
+            - v_contra_dxc
+            + u_contra_dyc[-1, 0, 0]
+            - u_contra_dyc
+        )
```

Comment:
@jdahm pointed out we can't have a write before read with offset here, will add a `u_contra_dyc_n1_0_0 = u_contra[-1, 0, 0] * dyc[-1, 0]` and use that instead of reading with offset.

Reply:
We also discussed potentially doing this by refactoring to a write-with-region before reading with an offset.